HW2/P4/P4.txt

```diff
@@ -0,0 +1,11 @@
+num_threads 1
+3.39166116714 seconds for 10 filter passes.
+
+num_threads 2
+1.74576187134 seconds for 10 filter passes.
+
+num_threads 4
+1.59977984428 seconds for 10 filter passes.
+
+With 2 threads the speedup is around 2 times relative to serial. However, with 4 threads, the run time is only slightly faster than 2 threads.
+In this implemention of the 3x3 median filter, we wait until iteration i finishes before starting iteration i + 1. Therefore, no threads start an iteration of filtering before the data for that iteration is ready. 
```

Comment:
It's nice that you get a speedup, but you're making all threads wait for each other at every iteration. Instead, what I did was have each thread run a for loop over all iterations, but I have thread N wait for a signal that threads N-1 and N+1 are done with the previous iteration that thread N was working on, to make sure that it's using up to date information. By doing that, a thread only waits for the two threads surrounding it.